Firestore/core/src/local/leveldb_index_manager.cc

```diff
@@ -536,21 +536,14 @@ IndexManager::IndexType LevelDbIndexManager::GetIndexType(
 
 absl::optional<std::vector<model::DocumentKey>>
 LevelDbIndexManager::GetDocumentsMatchingTarget(const core::Target& target) {
-  std::unordered_map<core::Target, model::FieldIndex> indexes;
+  std::vector<DocumentKey> result;
+  std::unordered_set<std::string> existing_keys;
   for (const auto& sub_target : GetSubTargets(target)) {
     auto index_opt = GetFieldIndex(sub_target);
     if (!index_opt.has_value()) {
       return absl::nullopt;
     }
-
-    indexes.insert({sub_target, index_opt.value()});
-  }
-
-  std::vector<DocumentKey> result;
-  std::unordered_set<std::string> existing_keys;
-  for (const auto& entry : indexes) {
-    const Target& sub_target = entry.first;
-    const FieldIndex& index = entry.second;
+    const FieldIndex& index = index_opt.value();
```

Comment:
Thanks for your explanation of why you made this change. It's true that it's now a deterministic order, but I think this has a negative performance impact.

The code essentially wants to achieve this: "if any of the disjunction terms doesn't have a client-side index, we cannot serve the query using the client-side index". 

Now imagine that you have a query like
`a==1 || b==2 || c==3 || d==4 || e==5`. And that we _do_ have a client-side index for `a` and `b` and `c` and `d` (but we don't have a client-side index for `e`).

The code now uses the indexes to perform the first 4 queries (including all the work on line 551 to 581) using the client-side index, and then when it reaches `e`, it realizes `e` doesn't have a client-side index and it'll have to throw away all the work it did for the first 4 branches.

In other words, "early exit" (returning nullopt) used to happen much earlier in such cases.

Reply:
Can we use a ordered_map instead?

Reply:
Thank you for pointing that out! I totally agree with your points. However `ordered_map` will compare the value of  `Target` to decide the sequence. So I choose to use `vector<pair<..., ...>>` to keep the order return from `GetSubTargets`.

Reply:
👍